src/t8_data/t8_element_array_iterator.hxx

```diff
@@ -0,0 +1,191 @@
+ * Since the actual data is serialized to char-bytes in the underlying array, we have to store an element pointer
+ *  \a dref_element_ resembling the actual (serialized) element withint the iterator class and cannot return a real
+ *  reference of the element pointer residing in the underlying array. Therefore, read-only operations are possible.
+ */
+class t8_element_array_iterator {
+ public:
+  using iterator_category = std::bidirectional_iterator_tag;
+  using difference_type = std::ptrdiff_t;
+  using pointer = t8_element_t**;
+  using value_type = t8_element_t*;
+  using reference = t8_element_t*&;
+
+  /* Constructors */
+  t8_element_array_iterator () = delete;
+  t8_element_array_iterator (const t8_element_array_t* element_array, const t8_locidx_t position)
+    : scheme_ { t8_element_array_get_scheme (element_array) }, array_ { t8_element_array_get_array (element_array) },
+      index_ { position } {};
+
+  /* Copy/Move Constructors/Assignment-Operators */
+  t8_element_array_iterator (const t8_element_array_iterator& other) = default;
+  t8_element_array_iterator&
+  operator= (const t8_element_array_iterator& other)
+    = default;
+  t8_element_array_iterator (t8_element_array_iterator&& other) = default;
+  t8_element_array_iterator&
+  operator= (t8_element_array_iterator&& other)
+    = default;
+
+  /* Destructor */
+  ~t8_element_array_iterator () = default;
+
+  /* Dereferencing operator of the iterator wrapper, returning a reference to const of the t8_element_t-pointer to the serialized bytes representing the actual serialized t8_element_t pointer. */
+  const reference
+  operator* ()
+  {
+    T8_ASSERT (index_ >= 0 && static_cast<size_t> (index_) < array_->elem_count);
+    dref_element_ = static_cast<t8_element_t*> (t8_sc_array_index_locidx (array_, index_));
+    return dref_element_;
+  };
+
+  /* Pre- and Postfix increment */
+  t8_element_array_iterator&
+  operator++ ()
+  {
+    ++index_;
+    return *this;
+  };
+  t8_element_array_iterator
+  operator++ (int)
+  {
+    t8_element_array_iterator tmp_iterator (*this);
+    ++(*this);
+    return tmp_iterator;
+  };
+
+  /* Pre- and Postfix decrement */
+  t8_element_array_iterator&
+  operator-- ()
+  {
+    --index_;
+    return *this;
+  };
+  t8_element_array_iterator
+  operator-- (int)
+  {
+    t8_element_array_iterator tmp_iterator (*this);
+    --(*this);
+    return tmp_iterator;
+  };
+
+  /* Return the index within the array the iterator currently points to [0,...,size]. */
+  t8_locidx_t
+  GetArrayIndex () const
+  {
+    return index_;
+  };
+
+  /* Compute the linear id at a given level for the element the iterator points to. */
+  t8_linearidx_t
+  GetLinearIDAtLevel (const int level)
+  {
+    T8_ASSERT (index_ >= 0 && static_cast<size_t> (index_) < array_->elem_count);
+    dref_element_ = static_cast<t8_element_t*> (t8_sc_array_index_locidx (array_, index_));
+    return scheme_->t8_element_get_linear_id (dref_element_, level);
```

Comment:
Does this work?
```suggestion
    return scheme_->t8_element_get_linear_id (*(*this), level);
```

Reply:
Yes, this would work and with the updated dereference operator it works nicely together